### PR TITLE
Mwpw 137866: Merch card 2.0

### PR DESCRIPTION
**Pull request by Axelcureno**
IN DRAFT STATE FOR INTERNAL REVIEW AT THE MOMENT. 

* Add your
* Specific
* Features or fixes

Resolves: [MWPW-NUMBER](https://jira.corp.adobe.com/browse/MWPW-NUMBER)

**Test URLs:**
- Before: https://main--milo--adobecom.hlx.page/?martech=off
- After: https://<branch>--milo--adobecom.hlx.page/?martech=off

### CODE REVIEW

File: libs/blocks/merch-card/img/ellipsis.svg
Comment (yesil) [inline libs/blocks/merch-card/img/ellipsis.svg:1]: how this file is used?

File: test/blocks/merch-card/merch-card.test.js [hunk 1]
```diff
@@ -6,59 +6,72 @@ const { default: init } = await import('../../../libs/blocks/merch-card/merch-ca
...
+    const heading = merchCard.querySelector('h3[slot="heading-xs"]');
+    const body = merchCard.querySelector('div[slot="body-xs"]');
+    const footer = merchCard.querySelector('div[slot="footer"]');
+    const buttons = footer.querySelectorAll('.con-button');
+    expect(merchCard).to.exist;
+    expect(body).to.exist;
+    expect(heading).to.exist;
+    expect(merchCard.getAttribute('variant')).to.be.equal('segment');
+    expect(heading.textContent).to.be.equal('Lorem ipsum dolor sit amet');
+    expect(body.textContent).to.be.equal('Maecenas porttitor congue massa. Fusce posuere, magna sed pulvinar ultricies, purus lectus malesuada libero, sit amet commodo magna eros quis urna. Nunc viverra imperdiet enim.See what\'s included | Learn more');
```
Comment (yesil) [inline test/blocks/merch-card/merch-card.test.js:20]: While we can generate descriptions/text easily with AI, I don't like using long form lorem ipsum texts. But is is trivial.

File: test/blocks/merch-card/merch-card.test.js [hunk 2]
```diff
@@ -153,6 +145,128 @@ describe('Merch Card', () => {
     });
   });
 
+  describe('Catalog Card', () => {
```
Comment (yesil) [inline test/blocks/merch-card/merch-card.test.js:148]: ❤️

**Review by yesil — COMMENTED** [review]
I like how you organised the tests. 
We might need to add checks on the analytics attributes as well.

File: libs/blocks/merch-card/img/ellipsis.svg
Comment (Axelcureno) [reply libs/blocks/merch-card/img/ellipsis.svg:1]: This SVG is used in the action menu of the catalog card.

Screenshot 2023-11-03 at 7 08 30 AM

File: libs/blocks/merch-card/img/secure-transaction.svg
[the remark this answers is not in the document]
Comment (Axelcureno) [reply libs/blocks/merch-card/img/secure-transaction.svg:1]: This SVG is used in the plans card, and will be also used in other variants not yet developed.

Screenshot 2023-11-03 at 7 10 12 AM

File: libs/blocks/merch-card/merch-card.js [hunk 3]
```diff
@@ -1,161 +1,155 @@
...
-  const links = pElement ? pElement.querySelectorAll('a') : el.querySelectorAll('a');
-  const list = el.querySelector('ul');
+const isHeadingTag = (tagName) => /^H[1-5]$/.test(tagName);
+const isParagraphTag = (tagName) => tagName === 'P';
+const isListTag = (tagName) => tagName === 'UL';
 
-  const inner = el.querySelector(':scope > div:not([class])');
-  inner.classList.add(`consonant-${cardType}-inner`);
+const createAndAppendTag = (tagName, attributes, content, parent) => {
+  const newElement = createTag(tagName, attributes, content);
```
Comment (JasonHowellSlavin) [inline libs/blocks/merch-card/merch-card.js:64]: `newElement` is a bit generic, maybe `newTag` ?

File: libs/blocks/merch-card/merch-card.js [hunk 4]
```diff
@@ -1,161 +1,155 @@
...
-    if (element.tagName.match(/^H[1-6]$/)) element.classList.add(`consonant-${cardType}-title`);
-    if (element.tagName.match(/^P$/)) element.classList.add(`consonant-${cardType}-description`);
-    if (element.tagName.match(/^UL$/)) {
-      list.classList.add(`consonant-${cardType}-list`);
-      list.querySelectorAll('li').forEach((li) => li.classList.add(`consonant-${cardType}-list-item`));
+    const { tagName } = element;
+    if (isHeadingTag(tagName)) {
+      createAndAppendTag(tagName, { slot: textStyles[tagName] }, element.innerHTML, merchCard);
+      return;
+    } if (isParagraphTag(tagName)) {
```
Comment (JasonHowellSlavin) [inline libs/blocks/merch-card/merch-card.js:78]: new line for this `if`

File: libs/blocks/merch-card/merch-card.js [hunk 5]
```diff
@@ -1,161 +1,155 @@
...
-  decorateFooter(el, altCta, cardType);
-  merchCard.append(inner);
+    if (isListTag(tagName)) {
+      const lastChildOfBody = bodySlot.lastElementChild;
+      if (lastChildOfBody) {
+        bodySlot.removeChild(lastChildOfBody);
+        element.prepend(lastChildOfBody);
+      }
+      createAndAppendTag('div', { slot: 'list' }, element, merchCard);
+    }
```
Comment (JasonHowellSlavin) [inline libs/blocks/merch-card/merch-card.js:90]: Could de nest I think
```suggestion
    if (bodySlot.lastElementChild) {
      bodySlot.removeChild(lastChildOfBody);
      element.prepend(lastChildOfBody);
    }
    
    if (isListTag(tagName)) {
      createAndAppendTag('div', { slot: 'list' }, element, merchCard);
    }
```

File: libs/blocks/merch-card/merch-card.js [hunk 6]
```diff
@@ -1,161 +1,155 @@
...
-  icons?.forEach((icon) => {
-    const url = icon.querySelector('img').src;
-    const iconDiv = createTag('div', { class: 'consonant-MerchCard-ProductIcon', style: `background-image: url(${url})` });
-    iconWrapper.appendChild(iconDiv);
-    icon.parentNode?.remove();
-  });
-  inner.prepend(iconWrapper);
+const getActionMenuContent = (el, ribbonMetadata) => {
+  const index = (ribbonMetadata !== null) ? 1 : 0;
+  const expectedChildren = (ribbonMetadata !== null) ? 3 : 2;
```
Comment (JasonHowellSlavin) [inline libs/blocks/merch-card/merch-card.js:109]: dont need `()` around ternary

File: libs/blocks/merch-card/merch-card.js [hunk 7]
```diff
@@ -1,161 +1,155 @@
...
-  section.classList.add('milo-card-section');
+  let section = el.closest('.section');
+  if (section.parentElement.classList.contains('fragment')) {
+    section.style.display = 'contents';
+    const fragment = section.parentElement;
+    fragment.style.display = 'contents';
+    const fragmentParent = fragment.parentElement;
+    fragmentParent.style.display = 'contents';
+    const parentSection = fragmentParent.parentElement;
+    section = parentSection;
```
Comment (JasonHowellSlavin) [inline libs/blocks/merch-card/merch-card.js:136]: Might be cleaner if you put all the `const`s at top, and all the style modifications below. Also, not sure why you reassign section on 136, I don't see it being used further in this file.

File: libs/blocks/merch-card/merch-card.js [hunk 8]
```diff
@@ -168,17 +162,60 @@ const init = (el) => {
...
-  if (innerCleanup.classList.length === 0) innerCleanup.remove();
+
+  const merchCard = createTag('merch-card', { class: el.className, variant: cardType });
+  if (ribbonMetadata !== null) {
+    const badge = returnRibbonStyle(ribbonMetadata);
+    if (badge !== null) {
+      merchCard.setAttribute('badge-background-color', badge.badgeBackgroundColor);
+      merchCard.setAttribute('badge-color', badge.badgeColor);
+      merchCard.setAttribute('badge-text', badge.badgeText);
+    }
```
Comment (JasonHowellSlavin) [inline libs/blocks/merch-card/merch-card.js:173]: Looks like you can un nest

```suggestion
  const badge = returnRibbonStyle(ribbonMetadata);
  if (badge !== null) {
    merchCard.setAttribute('badge-background-color', badge.badgeBackgroundColor);
    merchCard.setAttribute('badge-color', badge.badgeColor);
    merchCard.setAttribute('badge-text', badge.badgeText);
  }
```

File: libs/blocks/merch-card/merch-card.js [hunk 9]
```diff
@@ -168,17 +162,60 @@ const init = (el) => {
...
+  if (styles.includes('secure')) {
+    replaceKey('secure-transaction', getConfig())
+      .then((key) => merchCard.setAttribute('secure-label', key));
+  }
+  if (altCta) {
+    const label = checkBoxLabel(ctas, altCta);
+    if (label !== null) {
+      merchCard.setAttribute('checkbox-label', label);
+    }
+  }
```
Comment (JasonHowellSlavin) [inline libs/blocks/merch-card/merch-card.js:209]: Do a lot of checks in the instantiation, can de nest here too I think

```suggestion
  const label = checkBoxLabel(ctas, altCta);
  if (label !== null) {
    merchCard.setAttribute('checkbox-label', label);
  }
```

File: test/blocks/merch-card/merch-card.test.js [hunk 10]
```diff
@@ -6,59 +6,72 @@ const { default: init } = await import('../../../libs/blocks/merch-card/merch-ca
...
+    const buttons = footer.querySelectorAll('.con-button');
+    expect(merchCard).to.exist;
+    expect(body).to.exist;
+    expect(heading).to.exist;
+    expect(merchCard.getAttribute('variant')).to.be.equal('segment');
+    expect(heading.textContent).to.be.equal('Lorem ipsum dolor sit amet');
+    expect(body.textContent).to.be.equal('Maecenas porttitor congue massa. Fusce posuere, magna sed pulvinar ultricies, purus lectus malesuada libero, sit amet commodo magna eros quis urna. Nunc viverra imperdiet enim.See what\'s included | Learn more');
     expect(buttons.length).to.be.equal(2);
     expect(buttons[0].textContent).to.be.equal('Learn More');
     expect(buttons[1].textContent).to.be.equal('Save now');
```
Comment (JasonHowellSlavin) [inline test/blocks/merch-card/merch-card.test.js:23]: This many matchers in one test may make it hard to pinpoint an error should one occur. Also, if you need to ensure the structure of the final DOM element is exactly as you need it, you may consider a second "output" mock that you bring in and `deep.equal`.

File: test/blocks/merch-card/merch-card.test.js [hunk 11]
```diff
@@ -153,6 +145,128 @@ describe('Merch Card', () => {
...
+      expect(detail.textContent).to.be.equal('Desktop + Mobile');
+      expect(buttons.length).to.be.equal(2);
+      expect(buttons[0].textContent).to.be.equal('Learn More');
+      expect(buttons[1].textContent).to.be.equal('Save now');
+    });
+  });
+
   describe('UAR Card', () => {
     before(async () => {
       document.body.innerHTML = await readFile({ path: './mocks/uar-card.html' });
```
Comment (JasonHowellSlavin) [inline test/blocks/merch-card/merch-card.test.js:272]: A lot of these tests might be better served with snapshots. Not sure if we have implemented that or can in web test runner, but a shortcut would be have your input mock/output mock and deep equal.

**Pull request closed without merging**